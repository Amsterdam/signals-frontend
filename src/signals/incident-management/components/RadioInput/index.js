import React from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import { RadioGroup, Radio, Label as AscLabel, themeSpacing } from '@datapunt/asc-ui';

import InfoText from 'components/InfoText';

const Wrapper = styled.div`
  width: 100%;
  margin-bottom: ${themeSpacing(6)};
`;

const StyledLabel = styled(AscLabel)`
  * {
    font-weight: normal
  }
`;

const RadioInput = ({ name, display, values }) => {
  const Render = ({ handler, value: current }) => {
    let info;
    let label;
    const currentValue = values?.find(({ key }) => key === current);

    if (currentValue) {
      ({ info, value: label } = currentValue);
    }

    return (
      <Wrapper>
        <div className="mode_input text rij_verplicht">
          {display && <AscLabel htmlFor={`form${name}`} label={<strong>{display}</strong>} />}

          <RadioGroup name={name}>
            {values?.map(({ key, value }) => (
              <StyledLabel key={key} label={value}>
                <Radio
                  checked={current === key}
                  id={`${name}-${key}`}
                  data-testid={`${name}-${key}`}
                  {...handler('radio', key)}
                />
              </StyledLabel>
            ))}
          </RadioGroup>

          {info && <InfoText text={`${label}: ${info}`} />}
        </div>
      </Wrapper>
    );
  };

  Render.defaultProps = {
    touched: false,
  };

  Render.propTypes = {
    handler: PropTypes.func.isRequired,
    value: PropTypes.string,
    touched: PropTypes.bool,
  };

  return Render;
};

export default RadioInput;
