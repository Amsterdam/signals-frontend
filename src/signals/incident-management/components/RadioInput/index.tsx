// SPDX-License-Identifier: MPL-2.0
// Copyright (C) 2019 - 2022 Gemeente Amsterdam
import { RadioGroup, Label, themeSpacing } from '@amsterdam/asc-ui'
import type { ControllerRenderProps } from 'react-hook-form'
import styled from 'styled-components'

import InfoText from 'components/InfoText'
import Radio from 'components/RadioButton'

const Wrapper = styled.div`
  width: 100%;
  margin-bottom: ${themeSpacing(6)};
`

const StyledLabel = styled(Label)`
  align-self: baseline;

  * {
    font-weight: normal;
  }
`

type Props = {
  display: string
  values: [{ key: string; value: string; info: string }]
  sort?: boolean
} & ControllerRenderProps

const RadioInput = ({
  name,
  display,
  values,
  sort,
  value: current,
  ...props
}: Props) => {
  let info
  let label
  const currentValue = values?.find(({ key }) => key === current)
  if (currentValue) {
    ;({ info, value: label } = currentValue)
  }

  return (
    <Wrapper>
      <div className="mode_input text rij_verplicht">
        {display && (
          <Label htmlFor={`form${name}`} label={<strong>{display}</strong>} />
        )}
        <RadioGroup name={name}>
          {values?.map(({ key, value }) => (
            <StyledLabel key={key} label={value}>
              <Radio
                checked={current === key}
                id={`${name}-${key}`}
                value={key}
                data-testid={`${name}-${key}`}
                onChange={() => {
                  props.onChange(key)
                }}
              />
            </StyledLabel>
          ))}
        </RadioGroup>

        {info && <InfoText text={`${label}: ${info}`} className={''} />}
      </div>
    </Wrapper>
  )
}

export default RadioInput
