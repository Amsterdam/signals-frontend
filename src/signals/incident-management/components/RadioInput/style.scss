.radio-input {
  width: 100%;

  &__control {
    margin-bottom: 16px;
    margin-top: 12px;
  }
}
