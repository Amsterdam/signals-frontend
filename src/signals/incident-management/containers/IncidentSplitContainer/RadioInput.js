import React, { useState } from 'react';
import PropTypes from 'prop-types';

import { RadioGroup, Radio, Label } from '@datapunt/asc-ui';
import InfoText from 'components/InfoText';
import { StyledLabel, Wrapper } from './styled';

const getInfo = (options, value) => options.find(({ key: currentValue }) => currentValue === value);

const RadioInput = ({ id, name, display, options, initialValue, register }) => {
  const [selected, setSelected] = useState(getInfo(options, initialValue));

  return (
    <Wrapper>
      <div className="mode_input text rij_verplicht">
        <Label htmlFor={name} label={<strong>{display}</strong>} />

        <RadioGroup name={name}>
          {options.map(({ key, value }) => (
            <StyledLabel key={key} label={value}>
              <Radio
                checked={key === initialValue}
                id={`${id}-${key}`}
                value={key}
                ref={register}
                onChange={event => {
                  event.preventDefault();
                  setSelected(getInfo(options, event.target.value));
                }}
              />
            </StyledLabel>
          ))}
        </RadioGroup>

        {selected.info && <InfoText text={`${selected.value}: ${selected.info}`} />}
      </div>
    </Wrapper>
  );
};

RadioInput.propTypes = {
  id: PropTypes.string.isRequired,
  display: PropTypes.string.isRequired,
  initialValue: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  options: PropTypes.arrayOf(
    PropTypes.shape({
      key: PropTypes.string.isRequired,
      value: PropTypes.string.isRequired,
      info: PropTypes.string,
    })),
  register: PropTypes.func,
};

export default RadioInput;
