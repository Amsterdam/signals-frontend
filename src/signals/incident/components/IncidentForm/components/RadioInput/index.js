import React from 'react';
import PropTypes from 'prop-types';
import { map } from 'lodash';

import Header from '../Header/';

const RadioInput = ({ handler, touched, hasError, meta, parent, getError, validatorsOrOpts }) => (
  <div className={`${meta && meta.isVisible ? 'row' : ''}`}>
    {meta && meta.isVisible ?
      <div className={`${meta.className || 'col-12'} mode_input`}>
        <Header
          meta={meta}
          options={validatorsOrOpts}
          touched={touched}
          hasError={hasError}
          getError={getError}
        >
          <div className="antwoorden">
            {meta.values ? map(meta.values, (value, key) => (
              <div className="antwoord" key={`${meta.name}-${key}`}>
                <input
                  id={`${meta.name}-${key}`}
                  className="kenmerkradio"
                  {...handler('radio', value)}
                  onClick={(e) => parent.meta.updateIncident({ [meta.name]: e.target.value })}
                />
                <label htmlFor={`${meta.name}-${key}`}>{value}</label>
              </div>
            )) : ''}
          </div>
        </Header>
      </div>
       : ''}
  </div>
);

RadioInput.propTypes = {
  handler: PropTypes.func,
  touched: PropTypes.bool,
  getError: PropTypes.func.isRequired,
  hasError: PropTypes.func.isRequired,
  meta: PropTypes.object,
  parent: PropTypes.object,
  validatorsOrOpts: PropTypes.object
};

export default RadioInput;
