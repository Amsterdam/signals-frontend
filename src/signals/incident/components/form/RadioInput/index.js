import React, { useCallback } from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import { useDispatch } from 'react-redux';
import { Radio, Label } from '@datapunt/asc-ui';

import { resetExtraState, updateIncident } from 'signals/incident/containers/IncidentContainer/actions';

const StyledLabel = styled(Label)`
  & > * {
    font-weight: 400 !important;
  }
`;

const RadioInput = ({ checked, id, idAttr, label, info, name, resetsStateOnChange }) => {
  const dispatch = useDispatch();

  const onChange = useCallback(() => {
    if (resetsStateOnChange) {
      dispatch(resetExtraState());
    }

    dispatch(
      updateIncident({
        [name]: {
          id,
          label,
          info,
        },
      })
    );
  }, [dispatch, id, info, label, name, resetsStateOnChange]);

  return (
    <StyledLabel htmlFor={idAttr} label={label}>
      <Radio
        checked={checked}
        data-testid="inputUsingDispatch"
        id={idAttr}
        onChange={onChange}
        type="radio"
      />
    </StyledLabel>
  );
};

RadioInput.defaultProps = {
  checked: false,
  info: '',
  resetsStateOnChange: false,
};

RadioInput.propTypes = {
  checked: PropTypes.bool,
  id: PropTypes.string.isRequired,
  idAttr: PropTypes.string.isRequired,
  info: PropTypes.string,
  label: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  resetsStateOnChange: PropTypes.bool,
};

export default RadioInput;
