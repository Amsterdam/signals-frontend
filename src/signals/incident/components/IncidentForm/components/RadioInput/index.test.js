import React from 'react';
import { shallow } from 'enzyme';

import RadioInput from './index';

describe('Form component <RadioInput />', () => {
  const metaFields = {
    name: 'input-field-name',
    placeholder: 'type here',
    values: {
      foo: 'foo',
      bar: 'bar'
    }
  };
  let wrapper;
  let handler;
  let touched;
  let getError;
  let hasError;
  let parent;

  beforeEach(() => {
    handler = jest.fn();
    touched = false;
    getError = jest.fn();
    hasError = jest.fn();
    parent = {
      meta: {
        setIncident: jest.fn()
      }
    };

    wrapper = shallow(<RadioInput
      handler={handler}
      parent={parent}
      touched={touched}
      hasError={hasError}
      getError={getError}
    />);
  });

  describe('rendering', () => {
    it('should render radio fields correctly', () => {
      wrapper.setProps({
        meta: {
          ...metaFields,
          isVisible: true
        }
      });

      expect(handler).toHaveBeenCalledWith('radio', 'foo');
      expect(handler).toHaveBeenCalledWith('radio', 'bar');
      expect(wrapper).toMatchSnapshot();
    });

    it('should render zero radio fields when values are not supplied', () => {
      wrapper.setProps({
        meta: {
          ...metaFields,
          isVisible: true,
          values: undefined
        }
      });

      expect(handler).not.toHaveBeenCalled();
      expect(wrapper).toMatchSnapshot();
    });

    it('should render no radio field when not visible', () => {
      wrapper.setProps({
        meta: {
          ...metaFields,
          isVisible: false
        }
      });

      expect(handler).not.toHaveBeenCalled();
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('events', () => {
    const event = { target: { value: 'foo' } };

    it('sets incident when value changes', () => {
      wrapper.setProps({
        meta: {
          ...metaFields,
          isVisible: true,
          updateIncident: true
        }
      });

      wrapper.find('input').first().simulate('click', event);

      expect(parent.meta.setIncident).toHaveBeenCalledWith({
        'input-field-name': 'foo'
      });
    });

    it('does nothing when updateIncident is false', () => {
      wrapper.setProps({
        meta: {
          ...metaFields,
          isVisible: true,
          updateIncident: false
        }
      });

      wrapper.find('input').first().simulate('click', event);

      expect(parent.meta.setIncident).not.toHaveBeenCalled();
    });
  });
});
