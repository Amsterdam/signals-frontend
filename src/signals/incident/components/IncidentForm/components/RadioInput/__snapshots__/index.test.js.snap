// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form component <RadioInput /> rendering should render no radio field when not visible 1`] = `
<div
  className=""
/>
`;

exports[`Form component <RadioInput /> rendering should render radio fields correctly 1`] = `
<div
  className="row"
>
  <div
    className="col-12 mode_input"
  >
    <Header
      getError={[MockFunction]}
      hasError={[MockFunction]}
      meta={
        Object {
          "isVisible": true,
          "name": "input-field-name",
          "placeholder": "type here",
          "values": Object {
            "bar": "bar",
            "foo": "foo",
          },
        }
      }
      touched={false}
    >
      <div
        className="antwoorden"
      >
        <div
          className="antwoord"
          key="input-field-name-foo"
        >
          <input
            className="kenmerkradio"
            id="input-field-name-foo"
            onClick={[Function]}
          />
          <label
            htmlFor="input-field-name-foo"
          >
            foo
          </label>
        </div>
        <div
          className="antwoord"
          key="input-field-name-bar"
        >
          <input
            className="kenmerkradio"
            id="input-field-name-bar"
            onClick={[Function]}
          />
          <label
            htmlFor="input-field-name-bar"
          >
            bar
          </label>
        </div>
      </div>
    </Header>
  </div>
</div>
`;

exports[`Form component <RadioInput /> rendering should render zero radio fields when values are not supplied 1`] = `
<div
  className="row"
>
  <div
    className="col-12 mode_input"
  >
    <Header
      getError={[MockFunction]}
      hasError={[MockFunction]}
      meta={
        Object {
          "isVisible": true,
          "name": "input-field-name",
          "placeholder": "type here",
          "values": undefined,
        }
      }
      touched={false}
    >
      <div
        className="antwoorden"
      />
    </Header>
  </div>
</div>
`;
